### DISCUSSION TITLE
Modifying runAdjoint code to provide different forward pass solutions

### DISCUSSION THREAD

**Discussion by albisb** [mdolab/dafoam#557] [General]
Dear all,

I am investigating on a method called "semi-transient" adjoint gradient and I would like to ask for advice or feedback on the idea and about its feasibility using DAFoam. The idea of this method is to run a steady adjoint case (which DAFoam is capable to do) but on top of a frozen timestep of an unsteady flowfield (so, instead of starting from a steady forward pass, I would start from an unsteady forward pass). The methodology would be the following:
- Run an unsteady RANS using pimpleFoam. Store intermediate results which would be evaluated for the adjoint pass. For the sake of simplicity, let's just assume I store a single timestep.
- Run the steady adjoint (I assume it would still need to use pimpleFoam because of the first coloring step?) on top of the previous results.

Regarding how this could be done, I checked the runAdjoint code, and I have some questions, that I attach in the code below:
`
def runAdjoint(objFun=calcObjFuncValues, sensFun=calcObjFuncSens, fileName=None):
    """
    Just run the primal and adjoint
    """

    DASolver.runColoring() --> I understand this step as an initialization, that should be done with the same solver as the adjoint solution is expected.
    xDV = DVGeo.getValues()
    funcs = {}
    funcs, fail = objFun(xDV) --> Forward CFD pass. The objective is to compute the objective function values that are used later on for the adjoint pass. Does it really matter where these objective functions come from, given they are in the expected format?
    funcsSens = {}
    funcsSens, fail = sensFun(xDV, funcs) --> Using the objective function results, run the adjoint solution. Same solver as used in the runColoring step, but does it really need to use the same as the forward CFD pass?
`

Any feedback would be appreciated on the feasibiltiy of this process. I am not entirely sure if these modifications could be done successfully.

Best regards,

Àlex

**Comment by friedenhe** [comment]
The `objFun(xDV)` call will run the CFD and get the converged states (velocity, pressure, etc), and then use the converged states to compute objectives. Then, the following adjoint call will use the converged states (from memory) to run the adjoint. In other words, everything happens in the memory, the sensFun call will not read state variables from disk. If you want to do that, you have to modify the DAFoam source code and add a function to read the states manually and overwrite the existing ones in the memory. 

On a separated note, DAFoam supports time-accurate unsteady adjoints now. We are working on a paper and should have more detailed instructions/tutorials later this year.

**Reply by albisb** [reply]
Many thanks, I will take that into account for testing this possibility.
